pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -18,237 +18,351 @@ package garbagecollector
+	// GC caches the owners that do not exist according to the API server.
+	absentOwnerCache *UIDCache
 }
 
-func (ownerNode *node) addDependent(dependent *node) {
-	ownerNode.dependentsLock.Lock()
-	defer ownerNode.dependentsLock.Unlock()
-	ownerNode.dependents[dependent] = struct{}{}
-}
+func NewGarbageCollector(metaOnlyClientPool dynamic.ClientPool, clientPool dynamic.ClientPool, mapper meta.RESTMapper, deletableResources map[schema.GroupVersionResource]struct{}) (*GarbageCollector, error) {
+	attemptToDelete := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "garbage_collector_attempt_to_delete")
+	attemptToOrphan := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "garbage_collector_attempt_to_orphan")
+	absentOwnerCache := NewUIDCache(500)
+	gc := &GarbageCollector{
+		clientPool:            clientPool,
+		restMapper:            mapper,
+		attemptToDelete:       attemptToDelete,
+		attemptToOrphan:       attemptToOrphan,
+		registeredRateLimiter: NewRegisteredRateLimiter(deletableResources),
+		absentOwnerCache:      absentOwnerCache,
+	}
+	gb := &GraphBuilder{
+		metaOnlyClientPool:                  metaOnlyClientPool,
+		registeredRateLimiterForControllers: NewRegisteredRateLimiter(deletableResources),
+		restMapper:                          mapper,
+		graphChanges:                        workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "garbage_collector_graph_changes"),
+		uidToNode: &concurrentUIDToNode{
+			uidToNode: make(map[types.UID]*node),
+		},
+		attemptToDelete:  attemptToDelete,
+		attemptToOrphan:  attemptToOrphan,
+		absentOwnerCache: absentOwnerCache,
+	}
+	if err := gb.monitorsForResources(deletableResources); err != nil {
+		return nil, err
+	}
+	gc.dependencyGraphBuilder = gb
 
-func (ownerNode *node) deleteDependent(dependent *node) {
-	ownerNode.dependentsLock.Lock()
-	defer ownerNode.dependentsLock.Unlock()
-	delete(ownerNode.dependents, dependent)
+	return gc, nil
 }
 
-type eventType int
+func (gc *GarbageCollector) Run(workers int, stopCh <-chan struct{}) {
+	defer gc.attemptToDelete.ShutDown()
+	defer gc.attemptToOrphan.ShutDown()
+	defer gc.dependencyGraphBuilder.graphChanges.ShutDown()
 
-const (
-	addEvent eventType = iota
-	updateEvent
-	deleteEvent
-)
+	glog.Infof("Garbage Collector: Initializing")
+	gc.dependencyGraphBuilder.Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh, gc.dependencyGraphBuilder.HasSynced) {
+		return
+	}
+	glog.Infof("Garbage Collector: All resource monitors have synced. Proceeding to collect garbage")
 
-type event struct {
-	eventType eventType
-	obj       interface{}
-	// the update event comes with an old object, but it's not used by the garbage collector.
-	oldObj interface{}
+	// gc workers
+	for i := 0; i < workers; i++ {
+		go wait.Until(gc.runAttemptToDeleteWorker, 1*time.Second, stopCh)
+		go wait.Until(gc.runAttemptToOrphanWorker, 1*time.Second, stopCh)
```

Comment:
See this conversation: https://github.com/kubernetes/kubernetes/pull/38679#discussion_r92816746